rust-symcrypt/src/rsa.rs

```diff
@@ -0,0 +1,578 @@
+                key_usage.to_flag(),
+                rsa_key,
+            ) {
+                symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => Ok(RsaKeyPair {
+                    inner: rsa_key,
+                    key_usage: key_usage,
+                    number_format: num_format,
+                }),
+                err => Err(err.into()),
+            }
+        }
+    }
+
+    /// `size_of_primes()` returns a tuple containing the sizes, in bytes, of byte arrays large enough to store each of the two primes of the RSA key.
+    pub fn size_of_primes(&self) -> (u32, u32) {
+        unsafe {
+            // SAFETY: FFI calls
+            // Currently, only two prime RSA is supported, i.e. the only valid indexes are 0 and 1
+            let prime_1 = symcrypt_sys::SymCryptRsakeySizeofPrime(self.inner(), 0);
+            let prime_2 = symcrypt_sys::SymCryptRsakeySizeofPrime(self.inner(), 1);
+            (prime_1, prime_2)
+        }
+    }
+
+    /// `number_of_pub_primes()` returns a `u32` representing the number of primes associated with the [`RsaKeyPair`]
+    pub fn number_of_pub_primes(&self) -> u32 {
+        2
+        // !REVIEW: Can call SymCryptRsakeyGetNumberOfPrimes() instead, but it's a wasteful call since we know RsaKeyPair musth have value 2.
+        // this value is also assumed since this struct is RsaKeyPair
+    }
+
+    /// `export_key_pair_blob()` returns a [`RsaKeyPairBlob`] value.
+    pub fn export_key_pair_blob(&self) -> Result<RsaKeyPairBlob, SymCryptError> {
+        // Get size of primes only once
+        let (size_p, size_q) = self.size_of_primes();
+
+        // Allocate buffers for filling RsaKeyPairBlob
+        let mut modulus_buffer = vec![0u8; self.size_of_modulus() as usize];
+        let mut pub_exp = vec![0u64; 1];
+        let mut p = vec![0u8; size_p as usize];
+        let mut q = vec![0u8; size_q as usize];
+
+        let mut d_p = vec![0u8; size_p as usize];
+        let mut d_q = vec![0u8; size_q as usize];
+
+        let mut crt_coefficient = vec![0u8; size_p as usize];
+        let mut private_exponent = vec![0u8; self.size_of_modulus() as usize];
+
+        let mut primes_len = [
+            p.len() as symcrypt_sys::SIZE_T,
+            q.len() as symcrypt_sys::SIZE_T,
+        ];
+        let mut crt_lens = [
+            d_p.len() as symcrypt_sys::SIZE_T,
+            d_q.len() as symcrypt_sys::SIZE_T,
+        ];
+
+        unsafe {
+            // SAFETY: FFI calls
+            let result = symcrypt_sys::SymCryptRsakeyGetValue(
+                self.inner(),
+                modulus_buffer.as_mut_ptr(),
+                modulus_buffer.len() as symcrypt_sys::SIZE_T,
+                pub_exp.as_mut_ptr(),
+                1 as symcrypt_sys::UINT32,
+                [p.as_mut_ptr(), q.as_mut_ptr()].as_mut_ptr(),
+                primes_len.as_mut_ptr(),
+                2 as symcrypt_sys::UINT32,
+                self.number_format.to_num_format(),
```

Comment:
>           self.number_format.to_num_format(), [](http://example.com/codeflow?start=5&length=46)

If you want to support LSB, this should be an input parameter to the export. The caller could export multiple times using LSB or MSB.

Reply:
Resolved offline